### PR TITLE
Added support for ray binning for ray tracing in XR (case 1346374).

### PR DESCRIPTION
**Pull request by anisunity**
https://fogbugz.unity3d.com/f/cases/1346374/
Support ray binning for RTR and RTGI in performance. Visually should not change anything.

Testing status
Most test pass locally, we need to run yamato to check if the fails are local or not.

### CODE REVIEW

**Comment by remi-chapelain** [comment]
Launched the test on [yamato](https://yamato.cds.internal.unity3d.com/jobs/1485-Graphics/tree/HDRP%252Fray-binning-xr/.yamato%252Fall-hdrp_dxr.yml%2523HDRP_DXR_trunk/7525561/job).

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRenderPipeline.RaytracingDeferredLightLoop.cs [hunk 1]
```diff
@@ -328,9 +338,13 @@ static void RayMarchGBuffer(CommandBuffer cmd, in DeferredLightingRTRPassData da
...
                             ctx.cmd.SetRayTracingIntParam(data.parameters.gBufferRaytracingRT, HDShaderIDs._BufferSizeX, bufferSizeX);
 
+                            // Inject the data for the additional views
+                            ctx.cmd.SetRayTracingIntParams(data.parameters.gBufferRaytracingRT, HDShaderIDs._RayBinViewOffset, bufferSizeX * bufferSizeY);
+                            ctx.cmd.SetRayTracingIntParams(data.parameters.gBufferRaytracingRT, HDShaderIDs._RayBinTileViewOffset, numTilesRayBinX * numTilesRayBinY);
+
                             // A really nice tip is to dispatch the rays as a 1D array instead of 2D, the performance difference has been measured.
                             uint dispatchSize = (uint)(bufferSizeX * bufferSizeY);
-                            ctx.cmd.DispatchRays(data.parameters.gBufferRaytracingRT, m_RayGenGBufferBinned, dispatchSize, 1, 1);
+                            ctx.cmd.DispatchRays(data.parameters.gBufferRaytracingRT, m_RayGenGBufferBinned, dispatchSize, (uint)data.parameters.viewCount, 1);
```
Comment (fabien-unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRenderPipeline.RaytracingDeferredLightLoop.cs:347]: By convention we've been using the dispatch **Z** for the view count in all dispatch calls. Could you update this code ?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace [hunk 2]
```diff
@@ -176,6 +178,8 @@ void RayGenGBufferBinned()
     // Grab the dimensions of the current raytrace shader
     uint3 LaunchIndex = DispatchRaysIndex();
 
+    UNITY_XR_ASSIGN_VIEW_INDEX(LaunchIndex.y);
```
Comment (fabien-unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:181]: See previous comment about convention to use Z as view index.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace [hunk 3]
```diff
@@ -187,11 +191,11 @@ void RayGenGBufferBinned()
 
     // Read and unpack the coordiante to screen space coordinates
     uint globalBinIndex = tileIndex * RAY_BINNING_TILE_SIZE * RAY_BINNING_TILE_SIZE + localTileIndex;
-    uint packedPixelCoordinate = _RayBinResult[globalBinIndex];
+    uint packedPixelCoordinate = _RayBinResult[globalBinIndex + _RayBinViewOffset * LaunchIndex.y];
```
Comment (fabien-unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:194]: same here for Z

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace [hunk 4]
```diff
@@ -187,11 +191,11 @@ void RayGenGBufferBinned()
...
     uint2 currentPixelCoord = uint2((packedPixelCoordinate & 0xffff0000) >> 16, packedPixelCoordinate & 0xffff);
 
     // If the local index of this pixel is beyond the valid count, no need to compute it
-    if (localTileIndex > _RayBinSizeResult[tileIndex])
+    if (localTileIndex > _RayBinSizeResult[tileIndex + _RayBinTileViewOffset * LaunchIndex.y])
```
Comment (fabien-unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:198]: same here for Z

**Review by fabien-unity — APPROVED** [review]
LGTM
See small comment about using the Z dispatch index for the view index

**Comment by anisunity** [comment]
Review fixes done

**Review by sebastienlagarde — APPROVED** [review]

**Review by remi-chapelain — APPROVED** [review]
Don't have the hardware to test, trusting user feedbacks, automated tests and the rest of the reviewers ✔️

**Pull request merged**